src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -271,6 +288,32 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	protected void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needAttributes = getErrorChannel() != null
+				&& KafkaMessageDrivenChannelAdapter.this.retryTemplate == null;
+		if (needAttributes) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes || KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
```

Comment:
No; we can't do that; the recoverer might (or might not) actually recover the message.

If the recoverer throws an exception, we still need to publish to the error channel.